### PR TITLE
Ticket 1620/rollback event log scanner

### LINKED ISSUE #1620: IMA: getPastEvents sometimes work significantly slow as expected
**Description** 
Case: compatibility check 
IMA version 11/16: 2.0.0
IMA version 5/16: 1.5.0-beta.1
Skaled: the same for all - 3.16.1

**Step to reproduce**
Run transfers loop between schain
When processing message takes more than 60 seconds observe logs on working node

**Actual result** 
In some regular use case search of IMA msg can takes more then 30-... min, depends on amount of blocks on the schain. 
NOTE: Root cause investigate in progress 

**Expected result** 
IMA processing msg should not takes more than 40 seconds in average. getPastEvents should parse logs not longer than 5-7 seconds based on previous results, and results on nodes not affected by this problem.

NOTE: Look closer on logs, attention to PWA thread  on affected chain. 
Compare imaGPE_big.txt with imaIterativeScan13.txt and imaIterativeScan12.txt

[imaGPE_big.txt](https://github.com/skalenetwork/IMA/files/13022989/imaGPE_big.txt)
[imaGPE.txt](https://github.com/skalenetwork/IMA/files/13022986/imaGPE.txt)
[imaIterativeScan12.txt](https://github.com/skalenetwork/IMA/files/13022987/imaIterativeScan12.txt)
[imaIS12.txt](https://github.com/skalenetwork/IMA/files/13022988/imaIS12.txt)
[imaIterativeScan13.txt](https://github.com/skalenetwork/IMA/files/13022983/imaIterativeScan13.txt)
[imaGPE13.txt](https://github.com/skalenetwork/IMA/files/13022984/imaGPE13.txt)

```[tasklist]
### Tasks
- [ ] https://github.com/skalenetwork/IMA/issues/1554
- [ ] https://github.com/skalenetwork/IMA/issues/1554
```

```[tasklist]
### Tasks
- [ ] https://github.com/skalenetwork/IMA/issues/1554
```

### PR DESCRIPTION
**Pull request by sergiy-skalelabs**
Merge logging imporvements

### CODE REVIEW

File: npms/skale-ima/index.mjs [hunk 1]
```diff
@@ -51,9 +51,8 @@ async function findOutReferenceLogRecord(
     const bnMessageNumberToFind = owaspUtils.toBN( nMessageNumberToFind.toString() );
     const strEventName = "PreviousMessageReference";
     const arrLogRecords = await imaEventLogScan.safeGetPastEventsProgressive(
-        details, strLogPrefix,
-        ethersProvider, 10, joMessageProxy, strEventName,
-        bnBlockId, bnBlockId, joMessageProxy.filters[strEventName]()
+        details, strLogPrefix, ethersProvider, 10, joMessageProxy, strEventName,
+        bnBlockId, bnBlockId, joMessageProxy.filters[strEventName](), optsChainPair
```
Comment (DmytroNazarenko) [inline npms/skale-ima/index.mjs]: `optsChainPair` supposedly should be passed as a function parameter

File: npms/skale-ima/index.mjs [hunk 2]
```diff
@@ -432,11 +424,11 @@ async function gatherMessages( optsTransfer ) {
                 cc.notice( optsTransfer.strActionName ) + cc.debug( " for " ) +
                 cc.info( strEventName ) + cc.debug( " event..." ) + "\n" );
         }
-        r = await imaEventLogScan.safeGetPastEventsProgressive(
-            optsTransfer.details, optsTransfer.strLogPrefixShort, optsTransfer.ethersProviderSrc,
-            10, optsTransfer.joMessageProxySrc, strEventName, nBlockFrom, nBlockTo,
-            optsTransfer.joMessageProxySrc.filters[strEventName](
-                owaspUtils.ethersMod.ethers.utils.id( optsTransfer.chainNameDst ), // dstChainHash
+        r = await imaEventLogScan.safeGetPastEventsProgressive( optsTransfer.details,
```
Comment (DmytroNazarenko) [inline npms/skale-ima/index.mjs:426]: `optsTransfer.optsChainPair` should be passed as a last parameter

File: npms/skale-ima/index.mjs [hunk 3]
```diff
@@ -930,18 +922,14 @@ async function checkOutgoingMessageEvent( optsTransfer, joSChain ) {
...
-                        new owaspUtils.ethersMod.ethers.Contract(
-                            optsTransfer.imaState.chainProperties.sc
-                                .joAbiIMA.message_proxy_chain_address,
-                            optsTransfer.imaState.chainProperties.sc
-                                .joAbiIMA.message_proxy_chain_abi,
-                            ethersProviderNode
-                        );
+                    const joMessageProxyNode = new owaspUtils.ethersMod.ethers.Contract(
+                        sc.joAbiIMA.message_proxy_chain_address,
+                        sc.joAbiIMA.message_proxy_chain_abi,
```
Comment (DmytroNazarenko) [inline npms/skale-ima/index.mjs]: What is `sc`? Probably should be `optsTransfer.imaState.chainProperties.sc` instead?

File: npms/skale-ima/index.mjs [hunk 4]
```diff
@@ -930,18 +922,14 @@ async function checkOutgoingMessageEvent( optsTransfer, joSChain ) {
...
+                        ethersProviderNode );
                     const strEventName = "OutgoingMessage";
                     const node_r = await imaEventLogScan.safeGetPastEventsProgressive(
-                        optsTransfer.details, optsTransfer.strLogPrefixShort,
-                        ethersProviderNode, 10, joMessageProxyNode, strEventName,
+                        optsTransfer.details, optsTransfer.strLogPrefixShort, ethersProviderNode,
```
Comment (DmytroNazarenko) [inline npms/skale-ima/index.mjs:931]: `optsChainPair` supposedly should be passed as a function parameter

**Review by DmytroNazarenko — CHANGES_REQUESTED** [review]
Typos should be fixed

File: npms/skale-ima/index.mjs (hunk 1 above)
Comment (sergiy-skalelabs) [reply npms/skale-ima/index.mjs]: fixed

File: npms/skale-ima/index.mjs (hunk 3 above)
Comment (sergiy-skalelabs) [reply npms/skale-ima/index.mjs]: fixed

File: npms/skale-ima/index.mjs (hunk 4 above)
Comment (sergiy-skalelabs) [reply npms/skale-ima/index.mjs:931]: fixed

**Review by DmytroNazarenko — APPROVED** [review]

**Review by dmytrotkk — APPROVED** [review]

**Pull request merged**